linkml/generators/jsonschemagen.py

```diff
@@ -154,136 +205,110 @@ def extract_permissible_text(pv):
-                    prop["oneOf"] = ref_list
+                    prop = JsonSchema.ref_for(descendants)
                 else:
-                    prop = ref
+                    prop = JsonSchema.ref_for(reference)
         else:
             if slot.multivalued:
                 if reference is not None:
-                    prop = JsonObj(type="array", items={"$ref": reference})
+                    prop = JsonSchema.array_of(JsonSchema.ref_for(reference))
                 elif fmt is None:
-                    prop = JsonObj(type="array", items={"type": typ})
+                    prop = JsonSchema.array_of(JsonSchema({"type": typ}))
                 else:
-                    prop = JsonObj(type="array", items={"type": typ, "format": fmt})
+                    prop = JsonSchema.array_of(JsonSchema({"type": typ, "format": fmt}))
             else:
                 if reference is not None:
-                    prop = JsonObj({"$ref": reference})
+                    prop = JsonSchema.ref_for(reference)
                 elif fmt is None:
-                    prop = JsonObj(type=typ)
+                    prop = JsonSchema({"type": typ})
                 else:
-                    prop = JsonObj(type=typ, format=fmt)
+                    prop = JsonSchema({"type": typ, "format": fmt})
 
         if slot.description:
-            prop.description = slot.description
-        if slot.required:
-            self.clsobj.required.append(underscore(aliased_slot_name))
+            prop['description'] = slot.description
+
         if slot.pattern:
             # See https://github.com/linkml/linkml/issues/193
-            prop.pattern = slot.pattern
+            prop['pattern'] = slot.pattern
+
         if slot.minimum_value is not None:
-            prop.minimum = slot.minimum_value
+            prop['minimum'] = slot.minimum_value
+
         if slot.maximum_value is not None:
-            prop.maximum = slot.maximum_value
+            prop['maximum'] = slot.maximum_value
+
         if slot.equals_string is not None:
-            prop.const = slot.equals_string
+            prop['const'] = slot.equals_string
+
         if slot.equals_number is not None:
-            prop.const = slot.equals_number
-        self.clsobj.properties[underscore(aliased_slot_name)] = prop
+            prop['const'] = slot.equals_number
+
+        subschema.add_property(aliased_slot_name, prop, slot_is_required)
+
         if (
             self.top_class is not None and camelcase(self.top_class) == camelcase(cls.name)
         ) or (self.top_class is None and cls.tree_root):
-            self.schemaobj.properties[underscore(aliased_slot_name)] = prop
+            self.top_level_schema.add_property(aliased_slot_name, prop, slot_is_required)
+
+    def serialize(self, **kwargs) -> str:
+        self.start_schema(**kwargs)
+        for enum_definition in self.schemaview.all_enums().values():
+            self.visit_enum(enum_definition)
```

Comment:
it's maybe a bit odd keeping the method names as visit_X if it's not truly using the VP anymore... but YMMV

Reply:
I think you're right. I've started working on refactoring and adding methods for handling rules and I'll probably rename some existing methods as part of that.